llvm/lib/SYCLLowerIR/SYCLVirtualFunctionsAnalysis.cpp

```diff
@@ -59,32 +67,116 @@ void checkKernel(const Function *F, const CallGraphTy &CG) {
   checkKernelImpl(F, CG, CallStack);
 }
 
+void computeFunctionToKernelsMappingImpl(Function *Kernel, const Function *F,
+                                         const CallGraphTy &CG,
+                                         FuncToFuncMapTy &Mapping) {
+  CallGraphTy::const_iterator It = CG.find(F);
+  // It could be that the function itself is a leaf and doesn't call anything
+  if (It == CG.end())
+    return;
+
+  Mapping[F].insert(Kernel);
+
+  const SmallPtrSet<Value *, 8> &Callees = It->getSecond();
+  for (const Value *V : Callees)
+    if (auto *Callee = dyn_cast<Function>(V))
+      computeFunctionToKernelsMappingImpl(Kernel, Callee, CG, Mapping);
+}
+
+void computeFunctionToKernelsMapping(Function *Kernel, const CallGraphTy &CG,
+                                     FuncToFuncMapTy &Mapping) {
+  CallGraphTy::const_iterator It = CG.find(Kernel);
+  // It could be that kernel doesn't call anything
```

Comment:
```suggestion
  // It could be that the kernel doesn't call anything
```

Reply:
Applied in 9bef3a14ff2c8083378a19d7dd5a5c77d7ef1114